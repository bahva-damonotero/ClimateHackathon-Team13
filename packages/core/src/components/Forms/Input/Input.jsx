import React from 'react';
import PropTypes from 'prop-types';

import {
  createClassString,
  flattenChildren
} from '@boozallen-uip/uswds-react-utils';
import { FormError, Label } from '..';
import { validatorsToObject } from '../../util';

export default function Input({
  children,
  describedBy,
  error,
  id,
  inputWidth,
  onChange,
  value,
  success,
  type,
  validationId,
  validators,
  ...props
}) {
  const groupClasses = createClassString([
    error ? 'usa-input-group--error' : '',
    inputWidth ? `usa-input-group--${inputWidth}` : ''
  ]);
  const childComponents = flattenChildren(React.Children.toArray(children));
  const labelEls = [];
  let prefixComponent;
  let suffixComponent;

  for (const child of childComponents) {
    const componentType = child.type ? child.type.displayName : undefined;

    switch (componentType) {
      case 'InputPrefix':
        prefixComponent = child;
        break;

      case 'InputSuffix':
        suffixComponent = child;
        break;

      default:
        labelEls.push(child);
    }
  }

  const hasPrefixOrSuffix = prefixComponent || suffixComponent;

  const classes = createClassString([
    error ? 'usa-input--error' : '',
    success ? 'usa-input--success' : '',
    inputWidth && !hasPrefixOrSuffix ? `usa-input--${inputWidth}` : ''
  ]);

  const ariaDescribedBy =
    `${describedBy || ''} ${validationId || ''}`.trim() || undefined;

  const isTextArea = type === 'textarea';

  const inputEl = !isTextArea && (
    <input
      aria-describedby={ariaDescribedBy}
      className={createClassString(['usa-input', classes])}
      data-validation-element={validationId ? `#${validationId}` : undefined}
      id={id}
      name={id}
      onChange={onChange}
      type={type}
      value={value}
      {...(validators ? validatorsToObject(validators) : {})}
      {...props}
    />
  );

  return (
    <>
      <Label error={error} target={id}>
        {labelEls}
      </Label>

      <FormError error={error} />

      {isTextArea ? (
        <textarea
          aria-describedby={ariaDescribedBy}
          className={createClassString(['usa-textarea', classes])}
          id={id}
          name={id}
          onChange={onChange}
          value={value}
          {...props}
        />
      ) : (
        <>
          {hasPrefixOrSuffix ? (
            <div
              className={createClassString(['usa-input-group', groupClasses])}
            >
              {prefixComponent && React.cloneElement(prefixComponent)}
              {inputEl}
              {suffixComponent && React.cloneElement(suffixComponent)}
            </div>
          ) : (
            inputEl
          )}
        </>
      )}
    </>
  );
}

Input.propTypes = {
  /** Label for input */
  children: PropTypes.node.isRequired,
  /**
   * See [here](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute)
   */
  describedBy: PropTypes.string,
  /** Error message to display */
  error: PropTypes.string,
  /** ID of the input */
  id: PropTypes.string.isRequired,
  /**
   * Available modifiers to usa-input-group--[width] and usa-input--[width]
   * See [here](https://designsystem.digital.gov/components/input-prefix-suffix/#component-variants)
   */
  inputWidth: PropTypes.oneOf([
    '2xs',
    'xs',
    'sm',
    'small',
    'md',
    'medium',
    'lg',
    'xl',
    '2xl'
  ]),
  /** Function called when the input is modified */
  onChange: PropTypes.func,
  /** Used to set the value of the input */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** If `true`, the input is outlined in green */
  success: PropTypes.bool,
  /** Type of the input */
  type: PropTypes.oneOf(['number', 'password', 'text', 'textarea']),
  /** ID of the `<Validation />` associated with this input */
  validationId: PropTypes.string,
  /** Array of validators to be run on this input */
  validators: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      regex: PropTypes.string.isRequired
    })
  )
};

Input.defaultProps = {
  success: false,
  type: 'text'
};
