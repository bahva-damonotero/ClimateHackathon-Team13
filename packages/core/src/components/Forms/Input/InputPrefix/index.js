import InputPrefix from './InputPrefix';

export default InputPrefix;
