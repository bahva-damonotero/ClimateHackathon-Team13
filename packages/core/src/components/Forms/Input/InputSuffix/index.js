import InputSuffix from './InputSuffix';

export default InputSuffix;
