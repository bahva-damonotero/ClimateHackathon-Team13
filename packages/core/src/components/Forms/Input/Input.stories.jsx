import React from 'react';
import spriteSvg from '@uswds/uswds/img/sprite.svg';

import Input from '.';
import { Form, InputPrefix, InputSuffix } from '..';

export default {
  title: 'Forms/Input',
  component: Input
};

const Template = (args) => (
  <Form>
    <Input {...args} />
  </Form>
);

const creditCardSvg = (
  <svg role="img" focusable="false" className="usa-icon">
    <use
      xmlnsXlink="http://www.w3.org/1999/xlink"
      xlinkHref={`${spriteSvg}#credit_card`}
    />
  </svg>
);

const defaultArgs = {
  children: 'Label',
  id: 'input-example'
};

export const textInput = Template.bind({});
textInput.args = {
  ...defaultArgs
};

export const textArea = Template.bind({});
textArea.args = {
  ...defaultArgs,
  type: 'textarea'
};

export const passwordInput = Template.bind({});
passwordInput.args = {
  ...defaultArgs,
  type: 'password'
};

export const numberInput = Template.bind({});
numberInput.args = {
  ...defaultArgs,
  type: 'number'
};

export const disabled = Template.bind({});
disabled.args = {
  ...defaultArgs,
  disabled: true
};

export const error = Template.bind({});
error.args = {
  ...defaultArgs,
  error: 'Error with form field'
};

export const success = Template.bind({});
success.args = {
  ...defaultArgs,
  success: true
};

export const inputWidth = Template.bind({});
inputWidth.args = {
  ...defaultArgs,
  inputWidth: 'sm'
};

export const prefix = Template.bind({});
prefix.args = {
  ...defaultArgs,
  children: (
    <>
      <InputPrefix>{creditCardSvg}</InputPrefix>
      Cost
      <span className="usa-sr-only">Cost Input Prefixed with Dollar Sign</span>
    </>
  )
};

export const prefixWithError = Template.bind({});
prefixWithError.args = {
  ...defaultArgs,
  error: true,
  children: (
    <>
      <InputPrefix>{creditCardSvg}</InputPrefix>
      Cost
      <span className="usa-sr-only">Cost Input Prefixed with Dollar Sign</span>
    </>
  )
};

export const suffix = Template.bind({});
suffix.args = {
  ...defaultArgs,
  inputWidth: 'sm',
  children: (
    <>
      Weight, in pounds
      <InputSuffix>lbs.</InputSuffix>
    </>
  )
};

export const prefixAndSuffix = Template.bind({});
prefixAndSuffix.args = {
  ...defaultArgs,
  children: (
    <>
      <InputPrefix>$</InputPrefix>
      Total Cost
      <InputSuffix>.00</InputSuffix>
    </>
  )
};
