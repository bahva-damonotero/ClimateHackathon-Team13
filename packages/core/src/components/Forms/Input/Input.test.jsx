import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Input from '.';

describe('Input', () => {
  const id = 'input-example';
  const label = 'Input label';

  it('should render "success" variant', () => {
    const { container } = render(
      <Input id={id} readOnly success>
        {label}
      </Input>
    );

    expect(container.querySelector(`#${id}.usa-input--success`)).toBeTruthy();
  });

  it('should render with an error', () => {
    const { container } = render(
      <Input error="Invalid characters used" id={id} readOnly>
        {label}
      </Input>
    );

    expect(container.querySelector(`#${id}.usa-input--error`)).toBeTruthy();
  });

  it('should match snapshot (input)', () => {
    const tree = renderer
      .create(
        <Input id={id} readOnly>
          {label}
        </Input>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot (textarea)', () => {
    const tree = renderer
      .create(
        <Input id={id} readOnly type="textarea">
          {label}
        </Input>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
