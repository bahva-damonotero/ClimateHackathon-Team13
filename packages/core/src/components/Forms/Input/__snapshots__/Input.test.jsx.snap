// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot (input) 1`] = `
Array [
  <label
    className="usa-label"
    htmlFor="input-example"
  >
    Input label
  </label>,
  <input
    className="usa-input"
    id="input-example"
    name="input-example"
    readOnly={true}
    type="text"
  />,
]
`;

exports[`Input should match snapshot (textarea) 1`] = `
Array [
  <label
    className="usa-label"
    htmlFor="input-example"
  >
    Input label
  </label>,
  <textarea
    className="usa-textarea"
    id="input-example"
    name="input-example"
    readOnly={true}
  />,
]
`;
